File: chapter04/python/EditDistance.py


# 指针从后往前移动
def edit_distance(s1, s2):
    if len(s1) == 0:
        return len(s2)

    if len(s2) == 0:
        return len(s1)

    if s1[-1] == s2[-1]:
        return edit_distance(s1[:-1], s2[:-1])

    else:
        res = min(
            edit_distance(s1, s2[:-1]),        # 增
            edit_distance(s1[:-1], s2),                     # 删
            edit_distance(s1[:-1], s2[:-1])                # 替换
        )
        return res + 1


def edit_distance_iteration(s1, s2):
    import numpy as np
    n = len(s1)
    m = len(s2)
    dp_tale = np.zeros((n+1, m+1), dtype=np.int32)

    for i in range(n+1):
        dp_tale[i, 0] = i

    for i in range(m+1):
        dp_tale[0, i] = i
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i-1] == s2[j-1]:
                dp_tale[i, j] = dp_tale[i-1, j-1]
            else:
                res = min(
                    dp_tale[i - 1, j],
                    dp_tale[i, j - 1],
                    dp_tale[i - 1, j - 1],
                )
                dp_tale[i, j] = res + 1
    return dp_tale[n, m]


if __name__ == "__main__":
    s1 = "horse"
    s2 = "ros"
    print(edit_distance(s1, s2))
    print(edit_distance_iteration(s1, s2))

File: chapter04/cpp/EditDistance.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include "limits.h"

using namespace std;

// 指针从后往前移动
int edit_distance(const string& s1, const string& s2)
{   
    int m = s1.size();
    int n = s2.size();
    if(m == 0)
        return n;
    if(n == 0)
        return m;
    if(s1[m-1] == s2[n-1])
        return edit_distance(s1.substr(0, m-1), s2.substr(0, n-1));
    else
    {
        int res = min(
            {
                edit_distance(s1.substr(0, m), s2.substr(0, n-1)),        // 插入
                edit_distance(s1.substr(0, m-1), s2.substr(0, n)),        // 删除
                edit_distance(s1.substr(0, m-1), s2.substr(0, n-1)),      // 替换
            }
        );
        return res + 1;
    }
}

int edit_distance_iteration(const string& s1, const string& s2)
{
    int m = s1.size();
    int n = s2.size();
    vector<vector<int> > nums(n+1, vector<int>(m+1, 0));
    for(int i=0; i<=n; i++)
        nums[i][0] = i;
    
    for(int i=0; i<=m; i++)
        nums[0][i] = i;

    for(int i=1; i<=n; i++)
    {
        for(int j=1; j<=m; j++)
        {
            if(s2[i-1] == s1[j-1])
                nums[i][j] = nums[i-1][j-1];
            else
            {
                int res = min(
                    {
                        nums[i-1][j],
                        nums[i][j-1],
                        nums[i-1][j-1]
                    }
                );
                nums[i][j] = res + 1;
            }
        }
    }
    return nums[n][m];
}

int main()
{
    string s1 = "horse";
    string s2 = "ros";
    cout << edit_distance(s1, s2) << endl;
    cout << edit_distance_iteration(s1, s2) << endl;
    return 0;
}